### ISSUE TITLE
[salt-cloud] Linode deployment: Failed to create VM minion26. Configuration value 'dict' needs to be set

### ISSUE THREAD

**Issue by arnoldbechtoldt** [saltstack/salt#26349]
```
[ERROR   ] Failed to create VM minion26. Configuration value 'dict' needs to be set
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1256, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/linode.py", line 232, in create
    plan_id = get_plan_id(vm_['size'])
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/linode.py", line 679, in get_plan_id
    return avail_sizes()[label]['PLANID']
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/linode.py", line 148, in avail_sizes
    response = _query('avail', 'LinodePlans')
  File "/usr/lib/python2.7/site-packages/salt/cloud/clouds/linode.py", line 1147, in _query
    return result['dict']
KeyError: 'dict'
```

Salt cloud host:

```
# salt-call --versions
Salt Version:
           Salt: 2015.8.0rc1-153-g96c4fb6

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
         Python: 2.7.5 (default, Jun 24 2015, 00:41:19)
           RAET: Not Installed
        Tornado: 4.1
            ZMQ: 4.0.5
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1

System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-229.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

Cloud profiles/providers: 

https://github.com/bechtoldt/saltbox-vagrant/blob/master/shared/salt/devenv/master1/file_roots/pillar/salt-cloud.sls#L31

@rallytime FYI ^

Will try to investigate next week or so.

**Comment by rallytime** [comment]
Thanks @bechtoldt I'll take a look, if you don't beat me to it. You hit this just trying to create the VM, correct?

**Comment by bechtoldt** [comment]
@rallytime 

The root cause is very simple:

```
{'status': 503, 'error': 'HTTP 503: Service Temporarily Unavailable'}
```

:)

We need better A) error handling and B) better provider communication management.

A)
* retry API call n times after a failed API call?
* #23570
* #21474

B)
* don't fire off all HTTP requests at the same time (even when defining a smaller pool size)

**Comment by rallytime** [comment]
Oh, that error message helps a lot. I didn't see that in your first post and I couldn't get the stacktrace on my simple setup.

**Comment by rallytime** [comment]
When this issue gets addressed, it should be noted that the limits from Linode's API states that their call limits are 240 requests per 2 minute period, however, in reality, the limit is actually 2 calls per second.

**Comment by techhat** [comment]
@bechtoldt check out #29402.

**Comment by rmcintosh** [comment]
FWIW, it seems like the rate limiting implemented in #29402 doesn't seem to stop the 503s:

[ERROR   ] Failed to create VM min27. Configuration value 'dict' needs to be set

Verified that linode.py is up to date with the changes in that PR... not sure if maybe I'm missing something else, here...
